Add a driver for Mitsubishi Electric HVAC/airconditioners

dgwynne opened the pull request:
## Description:

Mitsubishi Electric (MiEl) aircons often have a serial port that can be communicated with to get their state, the temperature of the room they're installed in, and to control their operation. Supporting them in Tasmota is like how TuyaMCU devices work where the ESP handles the communication with the world, while the MCU/HVAC fiddles with the hardware.

This was inspired by the TuyaMCU support, and by [SwiCago/HeatPump](https://github.com/SwiCago/HeatPump). I've flashed everything else in the house with Tasmota, so I wanted to use it for this too.

So far this adds a bunch of commands for setting parameters on the aircon, and makes state and settings changes appear as MQTT messages. I haven't added support for tweaking things from the web interface yet though. However, it's Good Enough(tm) for me to be able to drive it with the MQTT HVAC integration in  Home-Assistant. The commands are:

- `HVACSetMode` configures the operating mode of the aircon, and takes `auto`, `heat`, `dry`, `cool`, and `fan_only` as arguments. Setting the mode does not affect the devices power state.
- `HVACSetHAMode` is like `HVACSetMode`, but works like how Home Assistant expects mode to work. Passing `off` as an argument turns the unit off, but setting any other mode implicitly turns the power on too.
- `HVACSetFan` sets the fan speed. It takes `auto`, `quiet`, `1`, `2`, `3`, and `4` as arguments.
- `HVACSetTemp` specifies the set point temperature/target temperature that the unit should try to reach when operating. It takes a number representing degrees centigrade from 16 to 31 inclusive.
- `HVACSetVane` sets the angle/mode of the vertical operation of the air vane. It takes `1` to `5`, `auto`, or `swing` as arguments.
- `HVACSetWideVane` sets the angle/mode of the horizontal operation of the air vane. It takes `LL` for far-left angle, `L` for left angle, `C` for center, `R` for right angle, `RR` for far-right angle, `split` for having air go both left and right, and `swing` for having it move around.

It also claims a `POWER` device slot, so you turn it on and off with `POWER 1` and `POWER 0`.

There's a bit of extra debug info still produced by the driver. It basically dumps the bytes it gets from the aircon in case there's extra information other models of the aircon can provide.

It adds the room temperature to the sensor output. This sensor is available even if the unit isn't operating, which is handy if you want to conditionally control the aircon. That looks like this:

```
15:59:57 MQT: tele/aircon_test/SENSOR = {"Time":"2020-09-05T15:59:57","MiElHVAC":{"Temperature":27.0,"status":"06000000000000000000000000000000","roomtemp":"03000011000000000000000000000000"},"TempUnit":"C"}
```

When settings are changed, or every teleperiod, it sends this:

```
15:59:57 MQT: tele/aircon_test/HVACSETTINGS = {"Power":"OFF","Mode":"fan_only","HAMode":"off","Temperature":26.0,"Fan":"1","Vane":"1","WideVane":"C","Bytes":"02000000070502010000030000000000"}
```

The code feels a bit different to everything else in the repo, because I have (a lot of) muscle memory for [KNF](https://man.openbsd.org/style.9). Hopefully it's not too bad to read/understand.

My current HA config for it looks like this:

```
sensor aircon_room_temp:
  - platform: mqtt
    name: "Loungeroom Temperature"
    device_class: temperature
    state_topic: "tele/aircon_test/SENSOR"
    unit_of_measurement: "C"
    value_template: "{{ value_json.MiElHVAC.Temperature }}"
    availability_topic: "tele/aircon_test/LWT"
    payload_available: "Online"
    payload_not_available: "Offline"

climate lounge:
  - platform: mqtt
    name: "Lounge"
    availability_topic: "tele/aircon_test/LWT"
    payload_available: "Online"
    payload_not_available: "Offline"
    current_temperature_topic: "tele/aircon_test/SENSOR"
    current_temperature_template: "{{ value_json.MiElHVAC.Temperature }}"
    precision: 1.0
    temperature_state_topic: "tele/aircon_test/HVACSETTINGS"
    temperature_state_template: "{{ value_json.Temperature }}"
    temperature_unit: "C"
    temp_step: 1
    min_temp: 16
    max_temp: 31
    temperature_command_topic: "cmnd/aircon_test/HVACSetTemp"
    mode_state_topic: "tele/aircon_test/HVACSETTINGS"
    mode_state_template: "{{ value_json.HAMode }}"
    modes:
      - "off"
      - "auto"
      - "heat"
      - "dry"
      - "cool"
      - "fan_only"
    mode_command_topic: "cmnd/aircon_test/HVACSetHAMode"
    fan_mode_state_topic: "tele/aircon_test/HVACSETTINGS"
    fan_mode_state_template: "{{ value_json.Fan }}"
    fan_modes:
      - "auto"
      - "1"
      - "2"
      - "3"
      - "4"
    fan_mode_command_topic: "cmnd/aircon_test/HVACSetFan"
```

**Related issue (if applicable):** fixes #<Tasmota issue number goes here>

## Checklist:
  - [x] The pull request is done against the latest dev branch
  - [x] Only relevant files were touched
  - [x] Only one feature/fix was added per PR.
  - [x] The code change is tested and works on Tasmota core ESP8266 V.2.7.4.1
  - [ ] The code change is tested and works on core ESP32 V.1.12.2
  - [x] I accept the [CLA](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md#contributor-license-agreement-cla).

_NOTE: The code change must pass CI tests. **Your PR cannot be merged unless tests pass**_

dgwynne committed:
provide names for MiEl HVAC pins

dgwynne committed:
a first pass at names for HVAC commands

dgwynne committed:
wire MiEl HVAC config up

dgwynne committed:
initial support for Mitsubishi Electric HVACs

based on SwiCago/HeatPump, but different because i wanted to understand
how it works and to try and fit it into tasmota properly.

dgwynne committed:
keep track of where im up to.

dgwynne committed:
this should be reasonably solid now.

dgwynne committed:
oops, should have tested this first

dgwynne committed:
stash the temperature bytes so we can see them change.

dgwynne committed:
respond with the value that was just set.

makes it consistent with other tasmota functionality, and is useful if
you want to see what HA is doing while on the console/mqtt sub

dgwynne committed:
add a home-assistant specific mode setting command.

HA thinks that "off" is a mode, while the actual hardware lets you
configure the mode separately to the power state.

dgwynne commented:
I didn't enable the driver in any config, so not sure the tests show anything except that I've tried to be cautious.

s-hadinger commented on tasmota/tasmota_template.h:
```diff
@@ -716,6 +719,10 @@ const uint8_t kGpioNiceList[] PROGMEM = {
   GPIO_TELEINFO_ENABLE,
 #endif
+#ifdef USE_MIEL_HVAC
+  GPIO_MIEL_HVAC_TX,
```
I'm afraid we can't afford new GPIOs right now as we are very close to the limit. Unless Theo has a magic way to extent the number of GPIOs, could you reuse existing ones?

dgwynne replied:
> I'm afraid we can't afford new GPIOs right now as we are very close to the limit. Unless Theo has a magic way to extent the number of GPIOs, could you reuse existing ones?

If using more GPIOs for another serial protocol is a problem, I think it should be solved more generally by the project. A useful reference point could be line disciplines on unixish systems. That's where a single serial driver always attaches to the hardware, and then the operator attaches a software driver called a line discipline that filters or takes over communication on the serial port.

Applying that here, Tasmota would only have the Serial Tx and Serial Rx GPIOs you could configure on physical pins. Then you'd deprecate others like the TuyaMCU Tx/Rx, ZigBee serial thing(?), and so on, but then add a config option to say what's on the other end of the serial port. The serial port would be like a virtual GPIO pin you attach a protocol to. Or you could rely on the module type to say how the serial port is used.

s-hadinger replied:
I will leave @arendst respond to that. Most of your code is compiled only if you enable the `#define USE_MIEL_HVAC` so whatever you include will not impact other users. For GPIOs it's different, since the GPIO numbers are reserved even if the user don't enable the feature. We indeed dedicate GPIOs for the most used features like Tuya and Zigbee.

s-hadinger reviewed (commented):
Thanks for your PR. Would it make sense to reuse the same syntax as IRHVAC commands and the same JSON attribute names, for the sake of consistency?

dgwynne commented:
> Thanks for your PR. Would it make sense to reuse the same syntax as IRHVAC commands and the same JSON attribute names, for the sake of consistency?

I happy to tweak the JSON attribute names in the info that gets published to be more consistent. I'd need some convincing that reusing the `IRHVAC` command and it's JSON argument is an improvement though.

dgwynne committed:
use the IRHVAC json key names.

dgwynne committed:
make the HVAC command names consistent with the JSON keys

dgwynne committed:
tweak the command macros so they match their strings.

dgwynne committed:
don't need a map for power on/off values

dgwynne committed:
the list of requests to make doesn't change.

dgwynne committed:
try to do a better job of connecting to the hvac.

don't rely on config to select the speed, but try iterating over 2400
and 9600 bps, and give it a tick before sending the command. only once
we get a reply to the connected message do we proceed with talking to
the hvac.

dgwynne committed:
let the connect "state machine" send the initial connect message

dgwynne committed:
a first pass at remote temperature setting support.

use the HVACRemoteTemperature command to pass a value in degC for the
current room temperature. pass "clear" to have the unit return to using
it's own sensor for room temperature readings.

dgwynne committed:
try to make the remotetemp handling a bit more robust.

be more generous with what is accepted, but massage the values before
encoding them for the hardware.

dgwynne committed:
fix the type on miel_hvac_remotetemp_degc2old().

dgwynne committed:
try and interpret the status bytes.

this also moves to only reporting status and room temperature via
SENSOR data. the raw bytes are still presented to try and help people
figure out what's going on.

dgwynne committed:
reorder SENSOR output so the temperature is before operation/compressor

the temp is almost certainly more interesting, so make it easier to see.

dgwynne committed:
factor out the msg input handling for roomtemp and status.

they both trigger sensor output and stash their bytes, so use common code
for that and pull the bytes apart when the sensor output is generated.

no functional change.

dgwynne committed:
if the hvac response to page 9 stuff, add it to sensor output.

dgwynne committed:
forgot to make sure the separator is set before printing stage info.

arendst commented:
As noted earlier I need to redesign the gpio selection code. It will be based on the current esp32 functionality so it's selection will be equal to the current method (select two serial gpios) but allows up to 2000 selections instead of the current 250.

I just need time to implement and keep legacy support.

dgwynne commented:
> I just need time to implement and keep legacy support.

I wish you luck. Anything I can tweak before this all falls out of my head?

matejzero commented:
I know this is not the best place to ask this, so if there is a better place, please let me know.

I would like to give this a try but I can't find any info on configuring Tasmota to work with my unit (I'm using ESP-01).

I was able to compile the Tasmota and under configuration, I selected `GPIO1` (serial out) as `MiEL HVAC TX` and `GPIO3` as `MiEL HVAC RX`. Is that all I need to do for tasmota to connect and receive data from HVAC?

dgwynne commented:
> ~I know this is not the best place to ask this, so if there is a better place, please let me know.~
> 
> ~I would like to give this a try but I can't find any info on configuring Tasmota to work with my unit (I'm using ESP-01).~
> 
> ~I was able to compile the Tasmota and under configuration, I selected `GPIO1` (serial out) as `MiEL HVAC TX` and `GPIO3` as `MiEL HVAC RX`. Is that all I need to do for tasmota to connect and receive data from HVAC?~
> 
> It turns out you only need to set RX / TX GPIOs and that's it. From the GUI, you only have an option to power on/off the HVAC, the rest is done via MQTT. Some more tweaking of the climate component and it's gonna be the bomb:)

I'm glad you go it working.

Could you tell me which model you have and what sensor and hvacstatus messages you get?

matejzero commented:
I have a MITSUBISHI ELECTRIC MSZ-AP35 model.

Sensor:
```json
{
    "Time": "2020-09-20T07:45:35",
    "MiElHVAC": {
        "Temperature": 26,
        "Operation": "ON",
        "Compressor": "OFF",
        "roomtemp": "030000100096B5B5FE42000173910000",
        "status": "0600000001000203B800004200000000",
        "stage": "09000000030000000000000000000000"
    },
    "TempUnit": "C"
}
```

HVAC status:
```json
{
    "Power": "ON",
    "Mode": "dry",
    "HAMode": "dry",
    "Temp": 25,
    "FanSpeed": "auto",
    "SwingV": "auto",
    "SwingH": "swing",
    "Bytes": "020000010206000000008CB232000000"
}
```

I noticed `Compressor` is always off, no matter if I'm in cooling or heating and hvacsettings temperature sometimes reset to 5C, but I didn't find a reason for it yet.

What is the difference between mode and hamode?
There are 2 power toggles in GUI (power1 and power2). Power1 turns on the HVAC, what does the other do?

dgwynne commented:
> I have a MITSUBISHI ELECTRIC MSZ-AP35 model.

You obviously get more info from a modern unit compared to mine.

> I noticed `Compressor` is always off, no matter if I'm in cooling or heating and hvacsettings temperature sometimes reset to 5C, but I didn't find a reason for it yet.

I believe I had that once during development when I was running the unit in dry mode, but not since then. I took a guess on what a bit meant when turning it into Compressor, but your unit has a lot of extra info which probably supersedes what mine does.

You would have to record a lot more of those sensor and status readings and correlate them to what you're observing the unit doing.

> What is the difference between mode and hamode?

"Mode" maps 100% to what the aircon unit is doing, where power state and the operating mode are presented as separate chunks of info. "HAMode", aka Home-Assistant Mode exists to cope with home-assistant's opinion on how things work. It's annoyingly limited in areas like this. HAMode adds Off as a selectable operating mode in addition to the ones the aircon actually supports.

> There are 2 power toggles in GUI (power1 and power2). Power1 turns on the HVAC, what does the other do?

I only have 1 power toggle at the moment. I had two when i set GPIO2 on the Wemos D1 up as an LED, and the second toggle controlled whether the led was on or off.

matejzero commented:
Thank for explaining this. I guessed HAMode is somehow connected with Home Assistant.

At the moment, the AC is doing everything I need with your firmware: mode, fan speed, swing V, swing H and setting the temperature. Maybe the extra options are for configuration of schedule, but I'm not interested in that right now.

As far as double power buttons go, this is how it looks.
![Screenshot 2020-09-22 at 08 05 32](https://user-images.githubusercontent.com/4689252/93848743-c223fe00-fcaa-11ea-9483-a387c7a41497.png)

My configuration:
![Screenshot 2020-09-22 at 08 05 39](https://user-images.githubusercontent.com/4689252/93848787-dcf67280-fcaa-11ea-9f38-503aa4bfb344.png)

dgwynne commented:
> Thank for explaining this. I guessed HAMode is somehow connected with Home Assistant.
> 
> At the moment, the AC is doing everything I need with your firmware: mode, fan speed, swing V, swing H and setting the temperature. Maybe the extra options are for configuration of schedule, but I'm not interested in that right now.

There's a completely separate page for time/schedule based stuff which I do not query at all. I'm pretty sure all the extra bytes you have a status info, which could be things like higher resolution temperature readings, humidity or how much power it's using and things like that.

> As far as double power buttons go, this is how it looks.

You're inheriting a button from using the Sonoff as a base module. If you switch to a Generic module with just the serial tx and rx set up in the template, you'll only get the single power toggle.

matejzero commented:
I'm noticing a weird issue. I set my target temperature to 25C and after random time, this gets reset to something else. Usually to some low temperature like 6C, sometimes 21C or 23C.

Are you also seeing this? I have set it to 25C now and will keep the console open to see if any special message comes in then the temperature is changes. Is there anything else I can do to debug this?

matejzero commented:
I happen to recreate the problem with console openen.

What I did was, set HVACSetTemp to 25, enable dry mode (had to do that twice for the AC to actually power on, but maybe I just didn't wait long enough) and then powered it off. After power off, HVACSETTINGS reported 7C instead of 25 as it should.

```json
# Set temperature
16:50:05 MQT: hvac/attic/tele/HVACSETTINGS = {"Power":"OFF","Mode":"dry","HAMode":"off","Temp":23.0,"FanSpeed":"auto","SwingV":"auto","SwingH":"swing","Bytes":"020000000208000000008CAE32000000"}
16:50:05 MQT: hvac/attic/stat/RESULT = {"HVACSetTemp":25}

16:50:09 MQT: hvac/attic/tele/SENSOR = {"Time":"2020-09-27T16:50:09","MiElHVAC":{"Temperature":25.0,"Operation":"OFF","Compressor":"OFF","roomtemp":"0300000F0001B3B3FE42000178C50000","status":"0600000000000103C200004200000000","stage":"09000000000000000000000000000000"},"TempUnit":"C"}
16:50:09 MQT: hvac/attic/tele/HVACSETTINGS = {"Power":"OFF","Mode":"dry","HAMode":"off","Temp":23.0,"FanSpeed":"auto","SwingV":"auto","SwingH":"swing","Bytes":"020000000208000000008CAE32000000"}
16:50:19 MQT: hvac/attic/tele/SENSOR = {"Time":"2020-09-27T16:50:19","MiElHVAC":{"Temperature":25.0,"Operation":"OFF","Compressor":"OFF","roomtemp":"0300000F0001B3B3FE42000178C50000","status":"0600000000000103C200004200000000","stage":"09000000000000000000000000000000"},"TempUnit":"C"}
16:50:19 MQT: hvac/attic/tele/HVACSETTINGS = {"Power":"OFF","Mode":"dry","HAMode":"off","Temp":23.0,"FanSpeed":"auto","SwingV":"auto","SwingH":"swing","Bytes":"020000000208000000008CAE32000000"}

# Start AC in dry mode
16:50:29 MQT: hvac/attic/stat/RESULT = {"HVACSetHAMode":"dry"}

16:50:29 MQT: hvac/attic/tele/SENSOR = {"Time":"2020-09-27T16:50:29","MiElHVAC":{"Temperature":25.0,"Operation":"OFF","Compressor":"OFF","roomtemp":"0300000F0001B3B3FE42000178C50000","status":"0600000000000103C200004200000000","stage":"09000000000000000000000000000000"},"TempUnit":"C"}
16:50:29 MQT: hvac/attic/tele/HVACSETTINGS = {"Power":"OFF","Mode":"dry","HAMode":"off","Temp":23.0,"FanSpeed":"auto","SwingV":"auto","SwingH":"swing","Bytes":"020000000208000000008CAE32000000"}
16:50:39 MQT: hvac/attic/tele/SENSOR = {"Time":"2020-09-27T16:50:39","MiElHVAC":{"Temperature":25.0,"Operation":"OFF","Compressor":"OFF","roomtemp":"0300000F0001B3B3FE42000178C50000","status":"0600000000000103C200004200000000","stage":"09000000000000000000000000000000"},"TempUnit":"C"}
16:50:39 MQT: hvac/attic/tele/HVACSETTINGS = {"Power":"OFF","Mode":"dry","HAMode":"off","Temp":23.0,"FanSpeed":"auto","SwingV":"auto","SwingH":"swing","Bytes":"020000000208000000008CAE32000000"}

# Start AC in dry mode again (after 15s)
16:50:45 MQT: hvac/attic/stat/RESULT = {"HVACSetHAMode":"dry"}
16:50:46 MQT: hvac/attic/stat/RESULT = {"POWER":"ON"}
16:50:46 MQT: hvac/attic/stat/POWER = ON

16:50:46 MQT: hvac/attic/tele/HVACSETTINGS = {"Power":"ON","Mode":"dry","HAMode":"dry","Temp":25.0,"FanSpeed":"auto","SwingV":"auto","SwingH":"swing","Bytes":"020000010206000000008CB232000000"}
16:50:47 MQT: hvac/attic/tele/SENSOR = {"Time":"2020-09-27T16:50:47","MiElHVAC":{"Temperature":25.0,"Operation":"OFF","Compressor":"OFF","roomtemp":"0300000F0001B3B3FE42000178C50000","status":"0600000000000103C200004200000000","stage":"09000000010000000000000000000000"},"TempUnit":"C"}
16:50:49 MQT: hvac/attic/tele/STATE = {"Time":"2020-09-27T16:50:49","Uptime":"2T22:48:46","UptimeSec":254926,"Heap":25,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER":"ON","Wifi":{"AP":1,"SSId":"wifiapname","BSSId":"74:4D:28:65:66:3A","Channel":6,"RSSI":78,"Signal":-61,"LinkCount":1,"Downtime":"0T00:00:03"}}
16:50:49 MQT: hvac/attic/tele/SENSOR = {"Time":"2020-09-27T16:50:49","MiElHVAC":{"Temperature":25.0,"Operation":"OFF","Compressor":"OFF","roomtemp":"0300000F0001B3B3FE42000178C50000","status":"0600000000000103C200004200000000","stage":"09000000010000000000000000000000"},"TempUnit":"C"}
16:50:49 MQT: hvac/attic/tele/HVACSETTINGS = {"Power":"ON","Mode":"dry","HAMode":"dry","Temp":25.0,"FanSpeed":"auto","SwingV":"auto","SwingH":"swing","Bytes":"020000010206000000008CB232000000"}
16:50:50 MQT: hvac/attic/tele/SENSOR = {"Time":"2020-09-27T16:50:50","MiElHVAC":{"Temperature":25.0,"Operation":"OFF","Compressor":"OFF","roomtemp":"0300000F0001B3B3FE42000178C60000","status":"0600000000000103C200004200000000","stage":"09000000010000000000000000000000"},"TempUnit":"C"}
16:50:58 MQT: hvac/attic/stat/RESULT = {"HVACSetHAMode":"off"}
16:50:59 MQT: hvac/attic/tele/STATE = {"Time":"2020-09-27T16:50:59","Uptime":"2T22:48:56","UptimeSec":254936,"Heap":25,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":26,"MqttCount":1,"POWER":"ON","Wifi":{"AP":1,"SSId":"wifiapname","BSSId":"74:4D:28:65:66:3A","Channel":6,"RSSI":76,"Signal":-62,"LinkCount":1,"Downtime":"0T00:00:03"}}
16:50:59 MQT: hvac/attic/tele/SENSOR = {"Time":"2020-09-27T16:50:59","MiElHVAC":{"Temperature":25.0,"Operation":"OFF","Compressor":"OFF","roomtemp":"0300000F0001B3B3FE42000178C60000","status":"0600000000000103C200004200000000","stage":"09000000010000000000000000000000"},"TempUnit":"C"}

# Power off the AC
16:50:59 MQT: hvac/attic/stat/RESULT = {"POWER":"OFF"}
16:50:59 MQT: hvac/attic/stat/POWER = OFF

# HVAC settings temperature changes to 7C
16:50:59 MQT: hvac/attic/tele/HVACSETTINGS = {"Power":"OFF","Mode":"dry","HAMode":"off","Temp":7.0,"FanSpeed":"auto","SwingV":"auto","SwingH":"swing","Bytes":"020000000218000000008CAF32000000"}
16:51:00 MQT: hvac/attic/tele/SENSOR = {"Time":"2020-09-27T16:51:00","MiElHVAC":{"Temperature":25.0,"Operation":"OFF","Compressor":"OFF","roomtemp":"0300000F0001B3B3FE42000178C60000","status":"0600000000000103C200004200000000","stage":"09000000000000000000000000000000"},"TempUnit":"C"}
```

dgwynne commented:
From what I can tell this is because the temperature setpoint doesn't mean anything when the aircon is in dry mode, so it basically forgets it. This isn't a problem if you're driving the aircon with an IR remote because the IR remote can't get state from the aircon, it just blindly sets all settings from what it's idea of the settings should be. If the IR remote goes from dry to cool/heat/auto, it just puts it's idea of the temperate back in place.

I'm not convinced the current behaviour is wrong, but I will agree it is annoying.

matejzero commented:
Oh that makes sense.

Let me do some tests with warm/cool mode and see if that sticks there.

eku commented on tasmota/xdrv_43_miel_hvac.ino:
```diff
@@ -0,0 +1,1287 @@
+
+#include <TasmotaSerial.h>
+
+#define XDRV_43			43
```
43 is already used by xdrv_43_mlx90640.ino

arendst commented:
Merged after needed adjustements. Thx.

mamrai1 commented:
is it included in precompiled binaries of 9 version?

jascdk commented:
Hi @dgwynne (David) 

Thanks for your integration. I have already set it up, and it works quite good so far:) I have one problem though. My unit is from Denmark. It has a mode called "iSafe". The purpose of this mode is just to let the unit go as low as 10 degrees. This mode is only (via the remote) accessible through the "heat" mode on the unit. Are you able to make the code so the temp can go te 10 degrees?

dgwynne commented:
> Hi @dgwynne (David)
> 
> Thanks for your integration. I have already set it up, and it works quite good so far:) I have one problem though. My unit is from Denmark. It has a mode called "iSafe". The purpose of this mode is just to let the unit go as low as 10 degrees. This mode is only (via the remote) accessible through the "heat" mode on the unit. Are you able to make the code so the temp can go te 10 degrees?

Hi jasck,

I'm happy to try, but I can't do much from here without help from you.

Can you try changing the iSafe mode on the remote, and seeing if the Bytes field in the HVACSETTINGS output changes? If it is changing, can you send through a copy of those lines along with some extra info about when you pressed the button? If we don't see bytes changing, we'll have to do some more in depth experimentation. 

More generally, is this the right place to work on this, or should we open another issue?

jascdk commented:
> > Hi @dgwynne (David)
> > Thanks for your integration. I have already set it up, and it works quite good so far:) I have one problem though. My unit is from Denmark. It has a mode called "iSafe". The purpose of this mode is just to let the unit go as low as 10 degrees. This mode is only (via the remote) accessible through the "heat" mode on the unit. Are you able to make the code so the temp can go te 10 degrees?
> 
> Hi jasck,
> 
> I'm happy to try, but I can't do much from here without help from you.
> 
> Can you try changing the iSafe mode on the remote, and seeing if the Bytes field in the HVACSETTINGS output changes? If it is changing, can you send through a copy of those lines along with some extra info about when you pressed the button? If we don't see bytes changing, we'll have to do some more in depth experimentation.
> 
> More generally, is this the right place to work on this, or should we open another issue?

Hi David.

Thanks for your response:) Well - at the moment I can't be at the place where the unit is (It is placed in our cottage). What I have done now is trying to force the temperature from 16 to 10 degrees remotely via HA and MQTT logs. It gives some different outputs in the bytes section. 

I don't know if you can use this - I will try to do the same but with the remote - perhaps in the next week or so.

Here is some short snippets of the MQTT log when going from 16 to 10. To do so I had to change the temp settings in the code.

```
Besked 1 er modtaget på tele/tasmota_mitsubishi/HVACSETTINGS kl. 21.43:
{
    "Power": "ON",
    "Mode": "heat",
    "HAMode": "heat",
    "Temp": 16,
    "FanSpeed": "auto",
    "SwingV": "swing",
    "SwingH": "C",
    "Bytes": "02000001010F0007000003A000000000"
}
 
 
Besked 2 er modtaget på tele/tasmota_mitsubishi/HVACSETTINGS kl. 21.43:
{
    "Power": "ON",
    "Mode": "heat",
    "HAMode": "heat",
    "Temp": 10,
    "FanSpeed": "auto",
    "SwingV": "swing",
    "SwingH": "C",
    "Bytes": "0200000101150007000003B500000000"
}
```
I don't know if this is the right place to work on this - im not so much of a novice here on GitHub. Should I open a specific issue in the "issues" section instead?

All the best

Jacob 

Denmark

jascdk commented:
Hi @dgwynne 

I now had the opportunity to come to our cottage.

I did som of the same reading like before, but I did the changes with the remote instead:

Her is what I got + some explanation:)

```
###################################################################################################################################### 
 
The "besked 10" danish word for "message 10" below is the lowest possible temperatur for normal operating. BUT on my model there is a "mode" called iSave which can be set by the remote. When the iSave is activated the next following json message comes through the console. The temperature remains on 16 degrees, but should drop to 10 degrees instead (which is the purpose for iSave). I can se the bytes is different for the two messages. It is important to note, that the iSave ONLY can be activated on the heat mode channel. It is not accesible on the other modes. 
 
I hope you can help:)
 
######################################################################################################################################
 
Besked 10 er modtaget på tele/tasmota_mitsubishi/HVACSETTINGS kl. 9.06:
{
    "Power": "ON",
    "Mode": "heat",
    "HAMode": "heat",
    "Temp": 16,
    "FanSpeed": "auto",
    "SwingV": "swing",
    "SwingH": "C",
    "Bytes": "02000001010F0007000003A000000000"
}
 
Besked 17 er modtaget på tele/tasmota_mitsubishi/HVACSETTINGS kl. 9.08:
{
    "Power": "ON",
    "Mode": "heat",
    "HAMode": "heat",
    "Temp": 16,
    "FanSpeed": "auto",
    "SwingV": "swing",
    "SwingH": "C",
    "Bytes": "02000001010F00070000039400000000"
}

```

logon84 commented:
Hi. Can someone test what temp is shown in HVACSETTINGS  when you set 25.5C in mitsubishi control panel? In my case I get 9C. 

Also would it possible to receive roomtemp with 0.5 precission? Yesterday I wrote an issue, couldn't find this thread to comment it here first.

mamrai1 commented:
You have to compile the firmwarw  on your own and add the driver

Smandurlo commented:
I am using this driver with my mitsubishi AC and it works like a charm on 5 internal units. Everything is perfect. Few days ago I read about esphome and its implementation for the mitsubishi ac. There is one very interesting thing: "It is possible to use an external temperature sensor to tell the heat pump what the room temperature is, rather than relying on its internal temperature sensor. You can do this by calling set_remote_temperature(float temp) on the mitsubishi_heatpump object in a lambda. Note that you can call set_remote_temperature(0) to switch back to the internal temperature sensor".

This is the website: https://github.com/geoffdavis/esphome-mitsubishiheatpump

Esphome has a very basic funcionality for vane, fan and hvac mode that doesn't fit my usage and cannot be changed.

Will it be possible to add it in tasmota? It is based on swicago, so it is probably already implemented in the code, but I don't know how to use it.

Thanks for any reply :-)

dgwynne commented:
> Will it be possible to add it in tasmota? It is based on swicago, so it is probably already implemented in the code, but I don't know how to use it.

It's already there, the command is `HVACRemoteTemp`. eg, `HVACRemoteTemp 26` will tell the AC that the temperature is 26 degrees and it will use that instead of its internal sensor. `HVACRemoteTemp clear` will tell the AC to use its own sensor again.

Smandurlo commented:
That is great! I will create a temperature sensor in HA that will write the value it in the mqtt cmd topic and clear it if the temperature sensor is not available. Is there a list of the commands available? I searched for it in the wiki and I didn't find anything.

Thank you very much :-)

Smandurlo commented:
I have a small problem: the HVACRemoteTemp allow only for int number: 33.5 that is a good value for the internal temperature sensor, is converted to 33. I only send to HVACRemoteTemp rounded number with 0.5 degree steps, so 33.7 is rounded to 33.5 but the AC use it as 33. If I write in the console 33.9 (not a valid number, but just to test it) the value is used as 33.

I think something in the code must be changed to allow 0.5 steps.
